### ISSUE TITLE
Require `if`s to have `{}`.

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#31448]
From the Flutter style guide (https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#separate-the-if-expression-from-its-statement):

```
// BAD:
if (notReady) return;

// GOOD:
if (notReady)
  return;

// ALSO GOOD:
if (notReady) {
  return;
}
```

I propose we change this to mark the middle case as BAD, i.e.:

```
// BAD:
if (notReady) return;

// BAD:
if (notReady)
  return;

// GOOD:
if (notReady) {
  return;
}
```

This is a frequent source of bugs, especially since we rely on manual formatting for indentation.

**Comment by dnfield** [comment]
IF we do this, we should update all instances in the repo to use this style.

**Comment by Hixie** [comment]
Why is it a frequent source of bugs?

**Comment by dnfield** [comment]
People add a second line under the if, or forget to indent, e.g.

```
if (notReady)
  makeMeReady();
  makeMeEvenMoreReady();
```

And it gets missed in code review.  Or, I end up seeing code sometimes like

```
if (notReady)
makeMeReady();
makeMeEvenMoreReady();
```

Which is also wrong and also sometimes get missed in review.

**Comment by Hixie** [comment]
If it's being missed in code review, then the code reviewer isn't being careful enough, IMHO.

Indeed, I would expect the code author themselves to catch this when self-reviewing the code before putting it up for review.

**Comment by dnfield** [comment]
In an ideal scenario this would not get missed - but I've definitely seen at least one or two cases recently where it was missed.  @jonahwilliams may have a couple as well, he was mentioning he saw some cases recently.

**Comment by jonahwilliams** [comment]
I've definitely caught this in code reviewer before, it is an easy mistake to make. I worry that I might not catch it again in the future. I'm wary of language choices that require reviewers/authors to be extra careful when there are safer options available.

**Comment by dnfield** [comment]
Here's an example that came up recently - it was authored by Jonah and reviewed by you and @tvolkert.  Everyone involved missed https://github.com/flutter/flutter/pull/15581/files#diff-afa141921d14a02769e479c13b0b775cR585

It was caught by @a14n in a formatting/cleanup PR and discussed here: https://github.com/flutter/flutter/pull/26921#discussion_r250147976

It wasn't exactly a bug, but it was pretty easy to miss and the source of some confusion later as to whether a bug had been introduced or the formatting had just been always off (it was that the formatting was just always off).  That could have been avoided if the style guide just required people to use braces.

**Comment by mraleph** [comment]
Note that this sort of (essentially formatting) errors would be impossible if Flutter Framework actually used `dartfmt`. 

I understand that there is certain amount of unhappiness with how it handles certain cases - but as Flutter team and external contributor community grows more and more cycles would be spend on just formatting code by hand and leaving nits about code formatting on PRs.

**Comment by Hixie** [comment]
Fixing dartfmt so it would be usable would definitely be the best solution here.

**Comment by Hixie** [comment]
Fundamentally the problem here is that requiring braces ends up making some code much less legible. There's cases where you just want to quickly bail, and having to use three lines instead of two makes the code overly verbose.

I think it's likely that in the example you cite above, both Todd and I were too complacent in our review. If we missed something that simple, what else might we have missed?

**Comment by dnfield** [comment]
Could we compromise on this being allowed:

```
if (returnEarly)
  return;
```

But this not?

```
if (someCondition)
  doSomething();
else
  doSomethingElse();
```

**Comment by christopherfujino** [comment]
@dnfield that makes sense, but it seems even more cognitively complex to remember...

**Comment by christopherfujino** [comment]
I'd like to go on record as saying I think the following is the most legible and very difficult to get wrong. I know a lot of people don't prefer it, but that's my 2 cents.

```
if (notReady) return;
```

**Comment by ccadieux** [comment]
I'll gladly take the verbose and safe option. Why choose an option that has historically caused a lot of bugs in software. Especially something so simple that a linter can catch the problem. 

Some supporting documentation:
https://www.imperialviolet.org/2014/02/22/applebug.html
https://dzone.com/articles/omitting-braces-not-just-a-mat
https://www.joelonsoftware.com/2005/05/11/making-wrong-code-look-wrong/

**Comment by Hixie** [comment]
@dnfield I'm happy to require braces if there's an `else`.

@christopherfujino The biggest problem with that pattern is that it makes it very hard to spot code flow when scanning the left margin.

**Comment by Hixie** [comment]
What we could do is have a lint that catches misindented code. A general rule could be that all statements in a block must have their first non-space character at the same left margin edge. That would catch the issue that we're worried about here.

cc @a14n in case that's something you'd like to add a lint for. :-)

**Comment by a14n** [comment]
>cc @a14n in case that's something you'd like to add a lint for. :-)

I could add such a lint but I'd prefer have a code formatter. A formatter is really helpful to catch issues. For instance in #30138 (which uses a dartfmt fork) the formatting result showed an [issue with precedence](https://github.com/flutter/flutter/pull/31486).

**Comment by mklim** [comment]
It's a pretty small bonus, but requiring braces in all cases also has the minor benefit of making patches that edit conditional blocks slightly more readable. There's extra noise in the diff if braces are only included when they're required.

```diff
 if (foo) {
   bar();
+  baz();
 }
```

vs

```diff
-if (foo)
+if (foo) {
   bar();
+  baz();
+}
```

**Comment by munificent** [comment]
> Fundamentally the problem here is that requiring braces ends up making some code much less legible. 

If that's true, why do Google's [Kotlin](https://developer.android.com/kotlin/style-guide#braces), [Java](https://google.github.io/styleguide/javaguide.html#s4.1-braces),  [JavaScript](https://google.github.io/styleguide/jsguide.html#formatting-braces), and [Objective-C](http://google.github.io/styleguide/objcguide.html#conditionals) style guides require braces on multi-line ifs?

Why do Go and Swift require braces for control flow in the language grammar?

**Comment by iapicca** [comment]
I'd like to offer my 2 cents

for `if` please allow this
```dart
if (_bool) => 'hi';
if (_bool) => _doSomething();
```

for `for` consistently with this
```dart
List<String> _list = [for (int i = 0; i < 10; i++) '$i'];
```
please allow this 
```dart
for (int i = 0; i < 10; i++) => '$i'];
for (int i = 0; i < 10; i++) =>  _doSomething();
```

for `while` please allow this
```dart
while (i < 10) =>=>  _doSomething();
```
and finally I'd love to
copying from `c# 8` 
allowing `switch` to this
```dart
String result = variable_expression switch
{
    case_1 => 'a',
    case_1 => 'b',
    _ => 'c' //default
};
```

**Comment by christopherfujino** [comment]
@Hixie isn't this issue about enabling the `curly_braces_in_flow_control_structures` lint throughout the framework?

**Comment by Hixie** [comment]
Yes, that was the original request. I don't think we should do that, it makes the code harder to read.
